include/swift/AST/Evaluator.h

```diff
@@ -69,6 +70,49 @@ class PrettyStackTraceRequest : public llvm::PrettyStackTraceEntry {
   }
 };
 
+/// An llvm::ErrorInfo container for a request in which a cycle was detected
+/// and diagnosed.
+template <typename Request>
+struct CyclicalRequestError : 
+  public llvm::ErrorInfo<CyclicalRequestError<Request>> {
+public:
+  static char ID;
+  const Request &request;
+
+  CyclicalRequestError(const Request &request): request(request) {}
+
+  virtual void log(llvm::raw_ostream &out) const {
```

Comment:
If we're going to log the error, it would be a lot nicer to use the evaluator's cycle-dumping machinery to show what the cycle is.